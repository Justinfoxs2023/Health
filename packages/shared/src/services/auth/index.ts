import { IUser } from '../../types';
import { STORAGE_KEYS } from '../../constants';
import { http } from '../http';
import { storage } from '../storage';

/** 登录参数 */
export interface ILoginParams {
  /** username 的描述 */
  username: string;
  /** password 的描述 */
  password: string;
  /** remember 的描述 */
  remember?: boolean;
}

/** 注册参数 */
export interface IRegisterParams {
  /** username 的描述 */
  username: string;
  /** password 的描述 */
  password: string;
  /** email 的描述 */
  email: string;
  /** phone 的描述 */
  phone?: string;
}

/** 认证响应 */
export interface IAuthResponse {
  /** user 的描述 */
  user: IUser;
  /** token 的描述 */
  token: string;
  /** refreshToken 的描述 */
  refreshToken: string;
}

/** 认证服务 */
class AuthService {
  private static instance: AuthService;
  private currentUser: IUser | null = null;
  private token: string | null = null;
  private refreshToken: string | null = null;
  private refreshPromise: Promise<void> | null = null;

  private constructor() {
    // 从存储中恢复认证状态
    this.token = storage.getItem(STORAGE_KEYS.TOKEN);
    this.refreshToken = storage.getItem(STORAGE_KEYS.REFRESH_TOKEN);
    this.currentUser = storage.getItem(STORAGE_KEYS.USER);

    // 如果有token，设置HTTP请求头
    if (this.token) {
      this.setAuthHeader(this.token);
    }

    // 添加响应拦截器处理token过期
    http.interceptors.response.use(
      response => response,
      async error => {
        const originalRequest = error.config;

        // 如果是401错误且不是刷新token的请求
        if (error.response?.status === 401 && !originalRequest._retry && this.refreshToken) {
          if (!this.refreshPromise) {
            this.refreshPromise = this.refreshAccessToken();
          }

          try {
            await this.refreshPromise;
            originalRequest._retry = true;
            return http(originalRequest);
          } finally {
            this.refreshPromise = null;
          }
        }

        return Promise.reject(error);
      },
    );
  }

  /** 获取单例 */
  public static getInstance(): AuthService {
    if (!AuthService.instance) {
      AuthService.instance = new AuthService();
    }
    return AuthService.instance;
  }

  /** 获取当前用户 */
  public getCurrentUser(): IUser | null {
    return this.currentUser;
  }

  /** 是否已认证 */
  public isAuthenticated(): boolean {
    return !!this.token && !!this.currentUser;
  }

  /** 登录 */
  public async login(params: ILoginParams): Promise<void> {
    const { data } = await http.post<IAuthResponse>('/auth/login', params);
    this.handleAuthSuccess(data);

    // 如果记住登录，保存刷新令牌
    if (params.remember) {
      storage.setItem(STORAGE_KEYS.REFRESH_TOKEN, data.refreshToken);
    }
  }

  /** 注册 */
  public async register(params: IRegisterParams): Promise<void> {
    const { data } = await http.post<IAuthResponse>('/auth/register', params);
    this.handleAuthSuccess(data);
  }

  /** 退出登录 */
  public async logout(): Promise<void> {
    if (this.token) {
      try {
        await http.post('/auth/logout');
      } catch (error) {
        console.error('Error in index.ts:', 'Logout failed:', error);
      }
    }

    this.clearAuth();
  }

  /** 刷新访问令牌 */
  private async refreshAccessToken(): Promise<void> {
    try {
      const { data } = await http.post<IAuthResponse>('/auth/refresh', {
        refreshToken: this.refreshToken,
      });

      this.token = data.token;
      this.setAuthHeader(data.token);
      storage.setItem(STORAGE_KEYS.TOKEN, data.token);
    } catch (error) {
      this.clearAuth();
      throw error;
    }
  }

  /** 处理认证成功 */
  private handleAuthSuccess(response: IAuthResponse): void {
    const { user, token, refreshToken } = response;

    this.currentUser = user;
    this.token = token;
    this.refreshToken = refreshToken;

    // 保存认证状态
    storage.setItem(STORAGE_KEYS.USER, user);
    storage.setItem(STORAGE_KEYS.TOKEN, token);

    // 设置请求头
    this.setAuthHeader(token);
  }

  /** 清理认证状态 */
  private clearAuth(): void {
    this.currentUser = null;
    this.token = null;
    this.refreshToken = null;

    // 清理存储
    storage.removeItem(STORAGE_KEYS.USER);
    storage.removeItem(STORAGE_KEYS.TOKEN);
    storage.removeItem(STORAGE_KEYS.REFRESH_TOKEN);

    // 清理请求头
    this.setAuthHeader(null);
  }

  /** 设置认证请求头 */
  private setAuthHeader(token: string | null): void {
    if (token) {
      http.defaults.headers.common.Authorization = `Bearer ${token}`;
    } else {
      delete http.defaults.headers.common.Authorization;
    }
  }
}

/** 认证服务实例 */
export const authService = AuthService.getInstance();
