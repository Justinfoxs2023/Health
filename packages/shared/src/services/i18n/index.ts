import { useState, useEffect } from 'react';

import { BehaviorSubject } from 'rxjs';

export type LocaleType = 'zh-CN' | 'en-US';

interface I18nConfig {
  /** dateFormat 的描述 */
  dateFormat: string;
  /** timeFormat 的描述 */
  timeFormat: string;
  /** numberFormat 的描述 */
  numberFormat: Record<LocaleType, Intl.NumberFormatOptions>;
  /** dateTimeFormat 的描述 */
  dateTimeFormat: Record<LocaleType, Intl.DateTimeFormatOptions>;
}

const defaultConfig: I18nConfig = {
  dateFormat: 'YYYY-MM-DD',
  timeFormat: 'HH:mm:ss',
  numberFormat: {
    'zh-CN': {
      style: 'decimal',
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    },
    'en-US': {
      style: 'decimal',
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    },
  },
  dateTimeFormat: {
    'zh-CN': {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      hour12: false,
    },
    'en-US': {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      hour12: true,
    },
  },
};

class I18nService {
  private locale$ = new BehaviorSubject<LocaleType>('zh-CN');
  private config$ = new BehaviorSubject<I18nConfig>(defaultConfig);
  private messages: Record<LocaleType, Record<string, string>> = {
    'zh-CN': {},
    'en-US': {},
  };

  constructor() {
    this.initI18n();
  }

  private initI18n() {
    // 从localStorage读取语言设置
    const savedLocale = localStorage.getItem('locale') as LocaleType;
    if (savedLocale) {
      this.setLocale(savedLocale);
    } else {
      // 使用浏览器默认语言
      const browserLocale = navigator.language;
      this.setLocale(browserLocale.startsWith('zh') ? 'zh-CN' : 'en-US');
    }

    // 从localStorage读取配置
    const savedConfig = localStorage.getItem('i18nConfig');
    if (savedConfig) {
      try {
        const config = JSON.parse(savedConfig);
        this.setConfig(config);
      } catch (error) {
        console.error('Error in index.ts:', 'Failed to parse i18n config:', error);
      }
    }
  }

  getLocale() {
    return this.locale$.value;
  }

  getConfig() {
    return this.config$.value;
  }

  setLocale(locale: LocaleType) {
    this.locale$.next(locale);
    localStorage.setItem('locale', locale);
    document.documentElement.setAttribute('lang', locale);

    // 触发自定义事件
    window.dispatchEvent(
      new CustomEvent('localechange', {
        detail: {
          locale: this.locale$.value,
          config: this.config$.value,
        },
      }),
    );
  }

  setConfig(config: Partial<I18nConfig>) {
    const newConfig = {
      ...this.config$.value,
      ...config,
    };
    this.config$.next(newConfig);
    localStorage.setItem('i18nConfig', JSON.stringify(newConfig));

    // 触发自定义事件
    window.dispatchEvent(
      new CustomEvent('localechange', {
        detail: {
          locale: this.locale$.value,
          config: newConfig,
        },
      }),
    );
  }

  async loadMessages(locale: LocaleType, messages: Record<string, string>) {
    this.messages[locale] = {
      ...this.messages[locale],
      ...messages,
    };
  }

  translate(key: string, params?: Record<string, string | number>) {
    const message = this.messages[this.locale$.value][key];
    if (!message) return key;

    if (!params) return message;

    return message.replace(/\{(\w+)\}/g, (_, key) => {
      return String(params[key] ?? `{${key}}`);
    });
  }

  formatNumber(value: number) {
    const locale = this.locale$.value;
    const options = this.config$.value.numberFormat[locale];
    return new Intl.NumberFormat(locale, options).format(value);
  }

  formatDate(date: Date | number) {
    const locale = this.locale$.value;
    const options = this.config$.value.dateTimeFormat[locale];
    return new Intl.DateTimeFormat(locale, options).format(date);
  }

  formatCurrency(value: number, currency = 'CNY') {
    const locale = this.locale$.value;
    return new Intl.NumberFormat(locale, {
      style: 'currency',
      currency,
    }).format(value);
  }

  formatRelativeTime(value: number, unit: Intl.RelativeTimeFormatUnit) {
    const locale = this.locale$.value;
    return new Intl.RelativeTimeFormat(locale, {
      numeric: 'auto',
    }).format(value, unit);
  }
}

export const i18n = new I18nService();

/** 国际化Hook */
export function useI18n(): {
  locale: any;
  config: any;
  t: (key: string, params?: Record<string, string | number> | undefined) => any;
  setLocale: (
    locale: import('D:/Health/packages/shared/src/services/i18n/index').LocaleType,
  ) => void;
  setConfig: (config: Partial<I18nConfig>) => void;
  formatNumber: (value: number) => string;
  formatDate: (date: number | Date) => string;
  formatCurrency: (value: number, currency?: string) => string;
  formatRelativeTime: (value: number, unit: Intl.RelativeTimeFormatUnit) => string;
} {
  const [locale, setLocale] = useState(i18n.getLocale());
  const [config, setConfig] = useState(i18n.getConfig());

  useEffect(() => {
    const handleLocaleChange = (e: CustomEvent) => {
      setLocale(e.detail.locale);
      setConfig(e.detail.config);
    };

    window.addEventListener('localechange', handleLocaleChange as EventListener);
    return () => {
      window.removeEventListener('localechange', handleLocaleChange as EventListener);
    };
  }, []);

  return {
    locale,
    config,
    t: i18n.translate.bind(i18n),
    setLocale: i18n.setLocale.bind(i18n),
    setConfig: i18n.setConfig.bind(i18n),
    formatNumber: i18n.formatNumber.bind(i18n),
    formatDate: i18n.formatDate.bind(i18n),
    formatCurrency: i18n.formatCurrency.bind(i18n),
    formatRelativeTime: i18n.formatRelativeTime.bind(i18n),
  };
}
