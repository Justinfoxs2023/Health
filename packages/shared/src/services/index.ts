// 导出核心服务
export * from './api';
export * from './auth';
export * from './storage';
export * from './notification';
export * from './healthData';
export * from './analytics';
export * from './logger';
export * from './http';
export * from './error';
export * from './theme';
export * from './i18n';
export * from './statistics';

// API服务基类
export class BaseApiService {
  protected baseURL: string;
  protected timeout: number;

  constructor() {
    this.baseURL = process.env.API_BASE_URL || 'http://localhost:3000';
    this.timeout = 10000;
  }

  protected handleError(error: any) {
    console.error('API Error:', error);
    throw error;
  }

  protected handleResponse(response: any) {
    if (!response.success) {
      throw new Error(response.message || '请求失败');
    }
    return response.data;
  }
}

// 健康数据服务
export class HealthDataService extends BaseApiService {
  async getHealthData(userId: string, type: string) {
    try {
      const response = await fetch(`${this.baseURL}/health-data/${userId}?type=${type}`);
      const data = await response.json();
      return this.handleResponse(data);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async saveHealthData(data: any) {
    try {
      const response = await fetch(`${this.baseURL}/health-data`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(data),
      });
      const result = await response.json();
      return this.handleResponse(result);
    } catch (error) {
      return this.handleError(error);
    }
  }
}

// 认证服务
export class AuthService extends BaseApiService {
  async login(username: string, password: string) {
    try {
      const response = await fetch(`${this.baseURL}/auth/login`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ username, password }),
      });
      const data = await response.json();
      return this.handleResponse(data);
    } catch (error) {
      return this.handleError(error);
    }
  }

  async logout() {
    try {
      const response = await fetch(`${this.baseURL}/auth/logout`, {
        method: 'POST',
      });
      const data = await response.json();
      return this.handleResponse(data);
    } catch (error) {
      return this.handleError(error);
    }
  }
}

// 存储服务
export class StorageService {
  setItem(key: string, value: any) {
    try {
      const serializedValue = JSON.stringify(value);
      localStorage.setItem(key, serializedValue);
    } catch (error) {
      console.error('Storage Error:', error);
    }
  }

  getItem(key: string) {
    try {
      const serializedValue = localStorage.getItem(key);
      return serializedValue ? JSON.parse(serializedValue) : null;
    } catch (error) {
      console.error('Storage Error:', error);
      return null;
    }
  }

  removeItem(key: string) {
    try {
      localStorage.removeItem(key);
    } catch (error) {
      console.error('Storage Error:', error);
    }
  }

  clear() {
    try {
      localStorage.clear();
    } catch (error) {
      console.error('Storage Error:', error);
    }
  }
} 