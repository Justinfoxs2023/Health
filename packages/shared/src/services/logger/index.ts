import { storage } from '../storage';

/** 日志级别 */
export type LogLevelType = 'debug' | 'info' | 'warn' | 'error';

/** 日志配置 */
export interface ILoggerConfig {
  /** 日志级别 */
  level: LogLevelType;
  /** 是否启用控制台输出 */
  enableConsole?: boolean;
  /** 是否启用存储 */
  enableStorage?: boolean;
  /** 存储保留时间（毫秒） */
  storageTTL?: number;
  /** 最大存储条数 */
  maxStorageEntries?: number;
  /** 自定义日志处理器 */
  customHandler?: (entry: ILogEntry) => void;
}

/** 日志条目 */
export interface ILogEntry {
  /** 时间戳 */
  timestamp: number;
  /** 日志级别 */
  level: LogLevelType;
  /** 消息 */
  message: string;
  /** 额外数据 */
  data?: any;
  /** 错误堆栈 */
  stack?: string;
  /** 来源 */
  source?: string;
}

const LOG_LEVELS: Record<LogLevelType, number> = {
  debug: 0,
  info: 1,
  warn: 2,
  error: 3,
};

/** 日志服务 */
class LoggerService {
  private static instance: LoggerService;
  private config: ILoggerConfig;
  private storageKey = 'logs';

  private constructor(config: ILoggerConfig) {
    this.config = {
      enableConsole: true,
      enableStorage: true,
      storageTTL: 7 * 24 * 60 * 60 * 1000, // 7天
      maxStorageEntries: 1000,
      ...config,
    };
  }

  public static getInstance(config: ILoggerConfig = { level: 'info' }): LoggerService {
    if (!LoggerService.instance) {
      LoggerService.instance = new LoggerService(config);
    }
    return LoggerService.instance;
  }

  /** 调试日志 */
  public debug(message: string, data?: any): void {
    this.log('debug', message, data);
  }

  /** 信息日志 */
  public info(message: string, data?: any): void {
    this.log('info', message, data);
  }

  /** 警告日志 */
  public warn(message: string, data?: any): void {
    this.log('warn', message, data);
  }

  /** 错误日志 */
  public error(message: string, error?: Error | any): void {
    this.log('error', message, error, error instanceof Error ? error.stack : undefined);
  }

  /** 获取所有日志 */
  public async getLogs(): Promise<ILogEntry[]> {
    if (!this.config.enableStorage) return [];
    return (await storage.getItem<ILogEntry[]>(this.storageKey)) || [];
  }

  /** 清空日志 */
  public async clearLogs(): Promise<void> {
    if (!this.config.enableStorage) return;
    await storage.removeItem(this.storageKey);
  }

  /** 设置配置 */
  public setConfig(config: Partial<ILoggerConfig>): void {
    this.config = {
      ...this.config,
      ...config,
    };
  }

  /** 获取配置 */
  public getConfig(): ILoggerConfig {
    return { ...this.config };
  }

  private async log(
    level: LogLevelType,
    message: string,
    data?: any,
    stack?: string,
  ): Promise<void> {
    if (LOG_LEVELS[level] < LOG_LEVELS[this.config.level]) return;

    const entry: ILogEntry = {
      timestamp: Date.now(),
      level,
      message,
      data,
      stack,
      source: this.getSource(),
    };

    // 控制台输出
    if (this.config.enableConsole) {
      this.consoleLog(entry);
    }

    // 存储日志
    if (this.config.enableStorage) {
      await this.storeLog(entry);
    }

    // 自定义处理
    if (this.config.customHandler) {
      this.config.customHandler(entry);
    }
  }

  private consoleLog(entry: ILogEntry): void {
    const { level, message, data, stack } = entry;
    const timestamp = new Date(entry.timestamp).toISOString();
    const prefix = `[${timestamp}] [${level.toUpperCase()}]`;

    switch (level) {
      case 'debug':
        console.debug(prefix, message, data);
        break;
      case 'info':
        console.info(prefix, message, data);
        break;
      case 'warn':
        console.warn(prefix, message, data);
        break;
      case 'error':
        console.error('Error in index.ts:', prefix, message, data);
        if (stack) console.error('Error in index.ts:', stack);
        break;
    }
  }

  private async storeLog(entry: ILogEntry): Promise<void> {
    const logs = await this.getLogs();
    logs.push(entry);

    // 限制日志条数
    if (logs.length > this.config.maxStorageEntries!) {
      logs.splice(0, logs.length - this.config.maxStorageEntries!);
    }

    // 移除过期日志
    const now = Date.now();
    const validLogs = logs.filter(log => now - log.timestamp < this.config.storageTTL!);

    await storage.setItem(this.storageKey, validLogs, {
      ttl: this.config.storageTTL,
    });
  }

  private getSource(): string {
    try {
      throw new Error();
    } catch (error) {
      const stack = (error as Error).stack || '';
      const frames = stack.split('\n');
      // 跳过前3帧（Error、getSource、log）
      const frame = frames[3] || '';
      const match = frame.match(/at\s+(.+?)\s+\((.+?)\)/) || frame.match(/at\s+(.+)$/);
      return match ? match[1] : 'unknown';
    }
  }
}

/** 日志服务实例 */
export const logger = LoggerService.getInstance();
