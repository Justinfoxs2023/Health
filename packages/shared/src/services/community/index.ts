import { logger } from '../logger';

/** 帖子类型 */
export enum PostType {
  EXPERIENCE = 'experience', // 经验分享
  QUESTION = 'question', // 问题咨询
  ACHIEVEMENT = 'achievement', // 成就展示
  ACTIVITY = 'activity', // 活动组织
  ARTICLE = 'article' // 专业文章
}

/** 帖子标签 */
export enum PostTag {
  DIET = 'diet', // 饮食
  EXERCISE = 'exercise', // 运动
  HEALTH = 'health', // 健康
  MEDICAL = 'medical', // 医疗
  MENTAL = 'mental', // 心理
  LIFESTYLE = 'lifestyle' // 生活方式
}

/** 帖子内容 */
export interface Post {
  /** ID */
  id: string;
  /** 作者ID */
  authorId: string;
  /** 标题 */
  title: string;
  /** 内容 */
  content: string;
  /** 类型 */
  type: PostType;
  /** 标签 */
  tags: PostTag[];
  /** 图片URL */
  images?: string[];
  /** 视频URL */
  video?: string;
  /** 点赞数 */
  likes: number;
  /** 评论数 */
  comments: number;
  /** 浏览数 */
  views: number;
  /** 创建时间 */
  createdAt: Date;
  /** 更新时间 */
  updatedAt: Date;
}

/** 评论内容 */
export interface Comment {
  /** ID */
  id: string;
  /** 帖子ID */
  postId: string;
  /** 作者ID */
  authorId: string;
  /** 内容 */
  content: string;
  /** 回复评论ID */
  replyTo?: string;
  /** 点赞数 */
  likes: number;
  /** 创建时间 */
  createdAt: Date;
  /** 更新时间 */
  updatedAt: Date;
}

/** 用户互动 */
export interface Interaction {
  /** 用户ID */
  userId: string;
  /** 帖子ID */
  postId?: string;
  /** 评论ID */
  commentId?: string;
  /** 互动类型 */
  type: 'like' | 'favorite' | 'share';
  /** 创建时间 */
  createdAt: Date;
}

/** 帖子查询选项 */
export interface PostQueryOptions {
  /** 类型 */
  type?: PostType;
  /** 标签 */
  tags?: PostTag[];
  /** 作者ID */
  authorId?: string;
  /** 排序方式 */
  sort?: 'latest' | 'popular' | 'recommended';
  /** 页码 */
  page?: number;
  /** 每页数量 */
  pageSize?: number;
}

/** 社区服务 */
export class CommunityService {
  private static instance: CommunityService;

  private constructor() {}

  /** 获取单例 */
  public static getInstance(): CommunityService {
    if (!CommunityService.instance) {
      CommunityService.instance = new CommunityService();
    }
    return CommunityService.instance;
  }

  /** 创建帖子 */
  public async createPost(
    authorId: string,
    data: {
      title: string;
      content: string;
      type: PostType;
      tags: PostTag[];
      images?: string[];
      video?: string;
    }
  ): Promise<Post> {
    try {
      const response = await fetch('/api/community/posts', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          authorId,
          ...data,
          createdAt: new Date(),
          updatedAt: new Date()
        })
      });

      if (!response.ok) {
        throw new Error('Failed to create post');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to create post', { error });
      throw error;
    }
  }

  /** 获取帖子列表 */
  public async getPosts(options: PostQueryOptions = {}): Promise<{
    posts: Post[];
    total: number;
  }> {
    try {
      const queryParams = new URLSearchParams();
      if (options.type) queryParams.append('type', options.type);
      if (options.tags) options.tags.forEach(tag => queryParams.append('tags', tag));
      if (options.authorId) queryParams.append('authorId', options.authorId);
      if (options.sort) queryParams.append('sort', options.sort);
      if (options.page) queryParams.append('page', options.page.toString());
      if (options.pageSize) queryParams.append('pageSize', options.pageSize.toString());

      const response = await fetch(`/api/community/posts?${queryParams}`);
      if (!response.ok) {
        throw new Error('Failed to fetch posts');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to fetch posts', { error });
      throw error;
    }
  }

  /** 获取帖子详情 */
  public async getPost(postId: string): Promise<Post> {
    try {
      const response = await fetch(`/api/community/posts/${postId}`);
      if (!response.ok) {
        throw new Error('Failed to fetch post');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to fetch post', { error });
      throw error;
    }
  }

  /** 更新帖子 */
  public async updatePost(
    postId: string,
    data: Partial<{
      title: string;
      content: string;
      type: PostType;
      tags: PostTag[];
      images: string[];
      video: string;
    }>
  ): Promise<Post> {
    try {
      const response = await fetch(`/api/community/posts/${postId}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          ...data,
          updatedAt: new Date()
        })
      });

      if (!response.ok) {
        throw new Error('Failed to update post');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to update post', { error });
      throw error;
    }
  }

  /** 删除帖子 */
  public async deletePost(postId: string): Promise<void> {
    try {
      const response = await fetch(`/api/community/posts/${postId}`, {
        method: 'DELETE'
      });

      if (!response.ok) {
        throw new Error('Failed to delete post');
      }
    } catch (error) {
      logger.error('Failed to delete post', { error });
      throw error;
    }
  }

  /** 创建评论 */
  public async createComment(
    authorId: string,
    data: {
      postId: string;
      content: string;
      replyTo?: string;
    }
  ): Promise<Comment> {
    try {
      const response = await fetch('/api/community/comments', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          authorId,
          ...data,
          createdAt: new Date(),
          updatedAt: new Date()
        })
      });

      if (!response.ok) {
        throw new Error('Failed to create comment');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to create comment', { error });
      throw error;
    }
  }

  /** 获取评论列表 */
  public async getComments(
    postId: string,
    options: {
      page?: number;
      pageSize?: number;
    } = {}
  ): Promise<{
    comments: Comment[];
    total: number;
  }> {
    try {
      const queryParams = new URLSearchParams();
      if (options.page) queryParams.append('page', options.page.toString());
      if (options.pageSize) queryParams.append('pageSize', options.pageSize.toString());

      const response = await fetch(`/api/community/posts/${postId}/comments?${queryParams}`);
      if (!response.ok) {
        throw new Error('Failed to fetch comments');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to fetch comments', { error });
      throw error;
    }
  }

  /** 更新评论 */
  public async updateComment(
    commentId: string,
    data: {
      content: string;
    }
  ): Promise<Comment> {
    try {
      const response = await fetch(`/api/community/comments/${commentId}`, {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          ...data,
          updatedAt: new Date()
        })
      });

      if (!response.ok) {
        throw new Error('Failed to update comment');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to update comment', { error });
      throw error;
    }
  }

  /** 删除评论 */
  public async deleteComment(commentId: string): Promise<void> {
    try {
      const response = await fetch(`/api/community/comments/${commentId}`, {
        method: 'DELETE'
      });

      if (!response.ok) {
        throw new Error('Failed to delete comment');
      }
    } catch (error) {
      logger.error('Failed to delete comment', { error });
      throw error;
    }
  }

  /** 添加互动 */
  public async addInteraction(
    userId: string,
    data: {
      postId?: string;
      commentId?: string;
      type: 'like' | 'favorite' | 'share';
    }
  ): Promise<Interaction> {
    try {
      const response = await fetch('/api/community/interactions', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          userId,
          ...data,
          createdAt: new Date()
        })
      });

      if (!response.ok) {
        throw new Error('Failed to add interaction');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to add interaction', { error });
      throw error;
    }
  }

  /** 删除互动 */
  public async removeInteraction(
    userId: string,
    data: {
      postId?: string;
      commentId?: string;
      type: 'like' | 'favorite' | 'share';
    }
  ): Promise<void> {
    try {
      const queryParams = new URLSearchParams();
      queryParams.append('userId', userId);
      if (data.postId) queryParams.append('postId', data.postId);
      if (data.commentId) queryParams.append('commentId', data.commentId);
      queryParams.append('type', data.type);

      const response = await fetch(`/api/community/interactions?${queryParams}`, {
        method: 'DELETE'
      });

      if (!response.ok) {
        throw new Error('Failed to remove interaction');
      }
    } catch (error) {
      logger.error('Failed to remove interaction', { error });
      throw error;
    }
  }

  /** 获取用户互动列表 */
  public async getUserInteractions(
    userId: string,
    type: 'like' | 'favorite' | 'share'
  ): Promise<{
    posts: Post[];
    comments: Comment[];
  }> {
    try {
      const response = await fetch(`/api/community/users/${userId}/interactions/${type}`);
      if (!response.ok) {
        throw new Error('Failed to fetch user interactions');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to fetch user interactions', { error });
      throw error;
    }
  }

  /** 搜索帖子 */
  public async searchPosts(
    keyword: string,
    options: PostQueryOptions = {}
  ): Promise<{
    posts: Post[];
    total: number;
  }> {
    try {
      const queryParams = new URLSearchParams();
      queryParams.append('keyword', keyword);
      if (options.type) queryParams.append('type', options.type);
      if (options.tags) options.tags.forEach(tag => queryParams.append('tags', tag));
      if (options.sort) queryParams.append('sort', options.sort);
      if (options.page) queryParams.append('page', options.page.toString());
      if (options.pageSize) queryParams.append('pageSize', options.pageSize.toString());

      const response = await fetch(`/api/community/search?${queryParams}`);
      if (!response.ok) {
        throw new Error('Failed to search posts');
      }

      return response.json();
    } catch (error) {
      logger.error('Failed to search posts', { error });
      throw error;
    }
  }
}

export const communityService = CommunityService.getInstance(); 