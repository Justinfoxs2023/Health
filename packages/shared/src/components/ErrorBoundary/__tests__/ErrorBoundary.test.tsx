import React from 'react';

import { ErrorBoundary } from '../index';
import { i18n } from '../../../services/i18n';
import { logger } from '../../../services/logger';
import { render, screen, fireEvent } from '@testing-library/react';

// 模拟日志服务
jest.mock('../../../services/logger', () => ({
  logger: {
    error: jest.fn(),
  },
}));

// 模拟国际化服务
jest.mock('../../../services/i18n', () => ({
  i18n: {
    t: (key: string) => key,
  },
  useI18n: () => ({
    t: (key: string) => key,
  }),
}));

describe('ErrorBoundary', () => {
  const ThrowError = () => {
    throw new Error('Test error');
  };

  beforeEach(() => {
    jest.clearAllMocks();
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('应该正常渲染子组件', () => {
    const { container } = render(
      <ErrorBoundary>
        <div>正常内容</div>
      </ErrorBoundary>,
    );

    expect(container).toHaveTextContent('正常内容');
  });

  it('应该捕获错误并显示错误UI', () => {
    const onError = jest.fn();

    render(
      <ErrorBoundary onError={onError}>
        <ThrowError />
      </ErrorBoundary>,
    );

    expect(screen.getByText('errorBoundary.title')).toBeInTheDocument();
    expect(screen.getByText('errorBoundary.message')).toBeInTheDocument();
    expect(screen.getByText(/Test error/)).toBeInTheDocument();
    expect(onError).toHaveBeenCalled();
    expect(logger.error).toHaveBeenCalledWith(
      'React error boundary caught error',
      expect.any(Object),
    );
  });

  it('应该使用自定义错误UI', () => {
    const fallback = <div>自定义错误UI</div>;

    render(
      <ErrorBoundary fallback={fallback}>
        <ThrowError />
      </ErrorBoundary>,
    );

    expect(screen.getByText('自定义错误UI')).toBeInTheDocument();
  });

  it('应该支持刷新页面', () => {
    const reloadSpy = jest.spyOn(window.location, 'reload').mockImplementation();

    render(
      <ErrorBoundary>
        <ThrowError />
      </ErrorBoundary>,
    );

    fireEvent.click(screen.getByText('errorBoundary.reload'));
    expect(reloadSpy).toHaveBeenCalled();
  });

  it('应该正确处理错误回调', () => {
    const onError = jest.fn();
    const error = new Error('Test error');

    render(
      <ErrorBoundary onError={onError}>
        <ThrowError />
      </ErrorBoundary>,
    );

    expect(onError).toHaveBeenCalledWith(
      error,
      expect.objectContaining({
        componentStack: expect.any(String),
      }),
    );
  });

  it('应该显示错误详情', () => {
    const error = new Error('Test error');
    error.stack = 'Error stack trace';

    const ThrowSpecificError = () => {
      throw error;
    };

    render(
      <ErrorBoundary>
        <ThrowSpecificError />
      </ErrorBoundary>,
    );

    expect(screen.getByText(/Test error/)).toBeInTheDocument();
    expect(screen.getByText(/Error stack trace/)).toBeInTheDocument();
  });

  it('应该应用正确的样式', () => {
    render(
      <ErrorBoundary>
        <ThrowError />
      </ErrorBoundary>,
    );

    const container = screen.getByText('errorBoundary.title').closest('.error-boundary');
    expect(container).toHaveClass('error-boundary');
    expect(container).toMatchSnapshot();
  });
});
