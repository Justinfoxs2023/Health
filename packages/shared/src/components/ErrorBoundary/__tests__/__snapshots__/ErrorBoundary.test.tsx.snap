// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary 应该应用正确的样式 1`] = `
<div
  class="error-boundary"
>
  <h2
    class="error-boundary__title"
  >
    errorBoundary.title
  </h2>
  <p
    class="error-boundary__message"
  >
    errorBoundary.message
  </p>
  <pre
    class="error-boundary__details"
  >
    Test error
    Error stack trace
  </pre>
  <button
    class="error-boundary__button"
  >
    errorBoundary.reload
  </button>
</div>
`; 